### ISSUE TITLE
Delayed cancel hits incorrect query.

### ISSUE THREAD

**Issue by skarsol** [pgbouncer/pgbouncer#245]
We're experiencing an issue where typically non-interactive queries (duration of milliseconds) are getting canceled with the message "canceling statement due to user request." There is no physical way for a user to insert a cancel there. Examining packet traces seems to indicate that pgbouncer is somehow delaying cancel requests and then when it ends up sending it on to postgres, a different query is now running on that backend and it gets cancelled.

In one example we have a front end server (S) running pgbouncer 1.7.2 (PGB1) that talks to a postgres server PSQL that is also running pgbouncer 1.7.2 (PGB2). 

S initiates a query that makes it to PSQL. After running for 8 seconds, S sends a cancel. PGB1 correctly propagates that cancel to the correct PID on PGB2. Almost 1 second later the result set makes its way from PSQL to S. No cancel is sent from PGB2 to PSQL.

8 minutes pass. 

PGB2 sends that pending cancel request to PSQL targeting the correct pid. That pid has processed thousands of queries in the interim. PSQL ACKs the cancel request, then that pid receives a new query, processes the query, returns the result set, receives yet another new query, and then cancels it out. The time frame between the cancelled query being sent to PSQL and the cancel response being generated is .000156 seconds.

![cancel](https://user-images.githubusercontent.com/1915152/32578433-d5d4a71c-c4a2-11e7-9d25-f59d5afbb06b.jpg)

**Comment by petere** [comment]
I haven't analyzed this in much detail, but my gut feeling is that that is just the way it works. The cancel request requires a second connection, and if that connection is being queued up in pgbouncer, then it could conceivably arrive much later when something else is already running. Cancellation works on the session, not a specific query.

Why does your client send so many cancel requests? Is that a part of your architecture?

**Comment by skarsol** [comment]
We have a fairly flexible interface that allows users to create varied queries across their data. If they manage to create a query that we haven't anticipated (or if they're just impatient) that runs slow then they'll hit cancel and make corrections to it (shrink date range, add other constraints, etc).

I understand the mechanics of how Postgres handles cancellations, but there should never be occasion for queries to queue more than a few seconds; our pool sizes are fairly generous. We're seeing pgbouncer sit on the cancel for multiple minutes before sending it. In the instant case above, pgbouncer sat on it for over 8 minutes. If that DB had been queuing for 8 minutes, alarms would have been going off. :)

The cancel above was sent from PGB to PSQL at 14:41:26. The cancel arrived at PGB at 14:33:52. Oodles of queries (not inside transactions) completed in that window.

At this point we've dropped the server lifetime down to 60 seconds which has greatly lowered the incident frequency (but it still occurs). We're going to change our application to just abandon queries and reconnect as opposed to trying to send an actual cancel until Postgres tries to be sane about cancels. (They've been uninterested in engaging with me on the postgres-bugs list so far). We'll have to see what the performance impact is of leaving those queries to run to completion.

I've tried to think of a way for pgbouncer to work around the issue, but have been unable to come up with anything that isn't susceptible to timing failures. It might be worthwhile to look at having cancel queries go into their own connection pool though in case that is what's going on here and we're just not seeing it.
